File: src/25-challenge/solution.py
from typing import Literal, Callable, cast

Instruction = Literal[">", "+", "-", "[", "]", "{", "}"]


def fn(code: str) -> int:
    result = 0
    position = 0

    def next_position():
        nonlocal position
        position += 1

    def increment():
        nonlocal result
        nonlocal position
        result += 1
        position += 1

    def decrement():
        nonlocal result
        nonlocal position
        result -= 1
        position += 1

    def loop_start():
        nonlocal position
        if not result:
            position = code.find("]", position)
        position += 1

    def loop_end():
        nonlocal position
        position = code.rfind("[", 0, position)

    def conditional_start():
        nonlocal position
        if not result:
            position = code.find("}", position)
        position += 1

    def conditional_end():
        nonlocal position
        position += 1

    instruction_actions: dict[Instruction, Callable[[], None]] = {
        ">": next_position,
        "+": increment,
        "-": decrement,
        "[": loop_start,
        "]": loop_end,
        "{": conditional_start,
        "}": conditional_end,
    }

    while position < len(code):
        instruction = cast(Instruction, code[position])
        instruction_actions[instruction]()

    return result

File: src/25-challenge/solution.ts
type Instruction = '>' | '+' | '-' | '[' | ']' | '{' | '}'

export default function fn(code: string): number {
	let result = 0
	let position = 0

	const instructionActions: Record<Instruction, () => void> = {
		'>': () => position++,
		'+': () => {
			result++
			position++
		},
		'-': () => {
			result--
			position++
		},
		'[': () => {
			if (!result) position = code.indexOf(']', position)
			position++
		},
		']': () => (position = code.lastIndexOf('[', position)),
		'{': () => {
			if (!result) position = code.indexOf('}', position)
			position++
		},
		'}': () => position++,
	}

	while (position < code.length) {
		const instruction = code[position] as Instruction
		instructionActions[instruction]()
	}

	return result
}
